Add chapter about localization in PHP (#2129)
TYPO3-Documentation/TYPO3CMS-Reference-CoreApi: Documentation/ExtensionArchitecture/HowTo/Localization/Php.rst

.. include:: /Includes.rst.txt
.. index::
   Localization; PHP
.. _extension-localization-php:

====================
Localization in PHP
====================

> Sometimes you have to localize a string in the PHP code, for

Reviewer: ```suggestion
Sometimes you have to localize a string in PHP code, for
```

example inside of a controller or a ViewHelper.

You can get the translation of a string into the current language via:

..  code-block:: php
    :caption: EXT:my_extension/Classes/SomeClass.php

>     $translatedString = $GLOBALS['LANG']->sL($translationKey);

Reviewer: $GLOBALS['LANG'] is mostly backend and cannot be used in frontend: When used in frontend and a backend user is logged in, the backend user language is used for lacalization.
See also: https://github.com/TYPO3/typo3/blob/main/typo3/sysext/core/Classes/Localization/LanguageService.php#L28

Mostly I use LocalizationUtility from Extbase. Don't know if this is best choice.

Author: However $GLOBALS['LANG'] is used by the translation ViewHelper in the Core, which in turn is used in the frontend, so I figured that this is best practice?

Reviewer: Strange, I really had the problem an hour ago, that GLOBALS['LANG'] used the backend user language (custom "gsw_CH") for frontend ("de" and "en"). After switching to LocalizationUtility it worked correctly.

Author: Maybe it is only available in the View and therefore for ViewHelpers. The class comments states "Some Cirumstances"

Reviewer: And I am getting an error when not logged-in into backend: Call to a member function sL() on null. I just use a plain user funtion for testing. Maybe translate viewhelper initialises something before using?

Reviewer: I remember vagely, that when the Admin Panel is activated, $GLOBALS['LANG'] works out of the box in frontend. That is obviously  then not the case without logged-in user.

Author: Here we also have the suggestion to use the Language Service: **https://docs.typo3.org/m/typo3/reference-coreapi/main/en-us/ApiOverview/Internationalization/XliffApi.html#label-access-in-php** However it does not state where we are at that moment :(

Reviewer: "In the backend" -> this is correct. But not in frontend context.

Reviewer: TSFE has a `sL` method which calls then ... tadaaa ... LanguageService:
https://github.com/TYPO3/typo3/blob/main/typo3/sysext/frontend/Classes/Controller/TypoScriptFrontendController.php#L2513

And the language service is initialized in TSFE the LanguageServiceFactory:
https://github.com/TYPO3/typo3/blob/main/typo3/sysext/frontend/Classes/Controller/TypoScriptFrontendController.php#L2524

This looks pretty good:
https://github.com/TYPO3/typo3/blob/main/typo3/sysext/core/Classes/Localization/LanguageServiceFactory.php#L46

But don't know if it is best practise to use that factory.

Reviewer: Maybe just use the TSFE. But TSFE is not injectable anymore since v12 (it has a state). In user functions you can use the available content object renderer, in controllers it should be also available without global variable.

Author: Would you like to take over? For me it is starting to be to complicate

Reviewer: I can make a draft for using TSFE in different contexts, but would like someone from core team to look over that before merging. Don't know if I can manage this in the next days. I'll try.

Then we can reference that chapter and suggest `TSFE->sL` for non-Extbase context and `LocalizationUtility` for Extbase context.

Where :php:`$translationKey` has the following format:

..  code-block:: php
    :caption: EXT:my_extension/Classes/SomeClass.php

    use TYPO3\CMS\Core\Utility\GeneralUtility;

>     $translationKey = 'LLL:EXT:' . GeneralUtility::camelCaseToLowerCaseUnderscored($extensionName)

Author: ```suggestion
    $translationKey = 'LLL:EXT:my_extension'
```

>     $translationKey = 'LLL:EXT:' . GeneralUtility::camelCaseToLowerCaseUnderscored($extensionName)

Reviewer: I find the usage of `camelCaseToLowerCaseUnderscored` overly complicated. In Extbase context I may have the extension name and can reuse it, mostly I have the extension key or just write it plain into the string.

If we go with this, we should link to the explanation of extension name/key, as many people don't know the difference, as I assume.

Author: Removed it

>     $translationKey = 'LLL:EXT:my_extension'
>         . '/Resources/Private/Language/locallang.xlf:' . $key;
> 

Author: ```suggestion
    $translationKey = 'LLL:EXT:my_extension' . '/Resources/Private/Language/locallang.xlf:' . $key;

```

As the access to this global variable will probably change in future versions
of TYPO3 it is considered best practice to use a wrapper function:

..  code-block:: php
    :caption: EXT:my_extension/Classes/SomeClass.php

    use TYPO3\CMS\Core\Localization\LanguageService;

    public function myMethod(string $translationKey): void {
        $translatedString = self::getLanguageService()->sL($translationKey);
        // ...
    }

>     protected static function getLanguageService(): LanguageService

Reviewer: Why static?

Author: Because no Class context is needed.

    {
        return $GLOBALS['LANG'];
    }

Localization in Extbase context
===============================

In Extbase context you can use the method
`\TYPO3\CMS\Extbase\Utility\LocalizationUtility::translate($key, $extensionName)`.

This method requires the localization key as the first and the extension's
name as the second parameter. Then the corresponding text in the current
language will be loaded from this extension's :file:`locallang.xlf` file.

> The function :php:`pi_getLL` takes translation overrides from TypoScript into

Author: ```suggestion
The function :php:`translate()` takes translation overrides from TypoScript into
```

> The function :php:`pi_getLL` takes translation overrides from TypoScript into

Reviewer: I assume you mean the `translate()` method instead `pi_getLL()`?
